src/wayland/text_input/text_input_handle.rs

```diff
@@ -186,13 +187,15 @@ impl TextInputHandle {
 #[derive(Debug)]
 pub struct TextInputUserData {
     pub(super) handle: TextInputHandle,
-    pub(crate) input_method_handle: InputMethodHandle,
+    pub(crate) input_method_handle: input_method::InputMethodHandle,
+    pub(crate) input_method_v3_handle: input_method_v3::InputMethodHandle,
```

Comment:
We shouldn't have two input methods at once. Including using different protocols. Maybe it would make sense if these were grouped together into a single enum?

Reply:
I don't think this works, because those are slots for input methods that may be filled in at any time. See a couple lines below:

```
if !data.input_method_handle.has_instance() && !data.input_method_v3_handle.has_instance() {
```

The instance for either may not be present, and because the `Option` is rather deeply nested (even behind a mutex each), there's no way to turn it into an enum.

I don't remember why it is so, but I do remember originally not liking it and exposing the Option directly. Eventually I just had to copy the same design in v3.

Reply:
Actually having two at once should be just fine, it depends on how it is used but functionally there really isn't any problem.

Reply:
I would even go so far as to say that several input method instances are a wanted feature.

Reply:
If you had to input methods clients though, won't both try to create a popup at the same time, and potentially insert text in response to the same input?

I don't know if there's really a way with the design of this protocol (any version) to have reasonable behavior with multiple input method engines running at once.

Reply:
That is correct, the first one acquires the input methos and locks the following ones.